Improvements to DiscusCalculateMultipleScatteringCorrection to support iterative process

DannyHindson opened the pull request:
**Description of work.**

Improvements to DiscusCalculateMultipleScatteringCorrection to support iterative process:
- performance improvements, especially if importance sampling is on:
  - the QS(Q) profile is prepared once per spectrum rather than once per bin
  - a histogram object called invPOfQ is now created once per spectrum and reused in each pass through a bin for loop rather than being created\destroyed on each pass through
- One small functional change to DiscusCalculateMultipleScatteringCorrection:
  - approach for converting S(Q) supplied as histogram to a points dataset has been modified to be consistent with flat interpolation used in algorithm
- extended the workspace creation helper:
  - expose some more parameters that control the bin sizes rather than hardcoding them (x0, deltax)
  - added an extra level to the "geographical" instrument definition (a ObjCompAssembly) so that the instrument can be saved to a Nexus file using SaveNexusESS. This means I can save instruments created by the helper without having to have an IDF file

Note the iterative process itself isn't part of this - that's a more "experimental" python script that I've not pushed yet

**To test:**

Run the following script. On the new code it completes in about 5 seconds for me on a Release build. It used to take about 1 min 30 secs (eg on the recently released 6.3 version of Mantid). Note - it should be quick because it uses a single scenario:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

test_ws = Load("PEARL00115245")
test_ws = ConvertUnits(test_ws, "Wavelength")
X=[5,6] # flat isotropic S(Q)
Y=[1,1] 
Sofq=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")
SetSample(InputWorkspace=test_ws,Geometry={'Shape': 'Cylinder', 'Height': 1.0,
                    'Radius': 0.2985, 'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'Ce-O2', 'MassDensity': 7.215})
DiscusMultipleScatteringCorrection(test_ws,Sofq, ImportanceSampling=True, NeutronPathsSingle=1, NeutronPathsMultiple=1,
                                   OutputWorkspace="MSResults")
```

Fixes #33049.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Various changes to support iterative MS calculation

Performance improvements in importance sampling functions:
- Skip prepareCumulativeProbForQ if not doing importance sampling
- Use vectors instead of histogram
- Use Histogram constructor without e in another place
- Use auto& instead of auto in a couple of places
- Reuse the same vectors in the bin loop to save allocating\freeing memory

Improve validation on S(Q) by checking for nans

Add CompAssembly to instrument helper function:
This means the instrument can be saved to a Nexus using
SaveNexusESS. SaveNexusESS should probably save the instrument
without requiring this but it seems like it requires the detectors
to be grouped into banks

Make algorithm more tolerant when given bad data - Check that wavelength
values are positive numbers

Modify the code which converts an input S(Q) profile from bin edges to points.
The original method was causing an offset in the corrected S(Q) profiles where
the peaks where moved to the right v slightly. Note - this caused one of the
unit tests to fail:
Need the spike to coincide with one of the detector two theta's for test to
work nicely and this has unfortunately moved it away

So move the spike in the test S(Q) to a new Q value. Note the most likely Q is still
slightly to the right of the S(Q) peak centre because the probabilty goes like
Q.S(Q) but close enough to the peak centre for the arithmetic to be valid
Also improved the windowing in the peak finding in the unit test

Make GetXMinMax work off bin centres rather than bin edges

Expose the start x and bin width in create2DWorkspaceWithGeographicalDetectors

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -174,6 +175,30 @@ std::map<std::string, std::string> DiscusMultipleScatteringCorrection::validateI
+
+  // determine the data range - only return min > 0. Bins with x=0 will be skipped later on
+  for (size_t workspaceIndex = 0; workspaceIndex < numberOfSpectra; workspaceIndex++) {
+    const auto &dataX = ws.points(workspaceIndex);
```
Do we need to add a check here fore is monitor/has detector?

DannyHindson replied:
I think those would be useful checks so I'll add them in

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -388,27 +427,21 @@ void DiscusMultipleScatteringCorrection::exec() {
-  if (qValues.back() < 2 * kinc) {
-    qValues.push_back(2 * kinc);
-    SQValues.push_back(SQValues.back());
+  if (qValues.back() < 2 * kmax) {
```
It doesn't look like there's a check for the last element being greater than kmax - is that not possible?

DannyHindson replied:
The qValues vector needs to extend at least as far as 2\*kmax for some later processing to work (the later processing is creation of a cumulative probability distribution). So if there is a value in qValues that is greater than 2*kmax that's fine, it just means we don't need to add an extra point in here.
Re the variable names - the user supplies two input workspaces: a) an S(Q) profile and b) an input sample workspace that defines the axes over which the multiple scattering calculation is to be calculated. "kmax" here is the largest k value in the input sample workspace while qValues is the list of q values in the S(Q) profile. Perhaps I'll just add a doxygen documentation comment to this function to explain because I'm struggling to come up with a snappy variable name. Does that sound OK?

RichardWaiteSTFC replied:
Ah OK I see, if you are making changes in the file already perhaps a comment/doc string would be helpful, if you don't mind

RichardWaiteSTFC commented on Framework/Algorithms/test/DiscusMultipleScatteringCorrectionTest.h:
```diff
@@ -95,28 +95,32 @@ class DiscusMultipleScatteringCorrectionTest : public CxxTest::TestSuite {
     if (peakPos.size() > 0) {
       TS_ASSERT_EQUALS(peakPos.front(), 0);
-      TS_ASSERT((static_cast<double>(peakPos.back()) * 0.2 > 93) && (static_cast<double>(peakPos.back()) * 0.2 < 94));
+      TS_ASSERT((static_cast<double>(peakPos.back()) * 0.2 >= 114) &&
```
Nice addition to the test, where does the factor of 0.2 come from, might be worth a comment?

DannyHindson replied:
The factor converts a workspace index to a two theta value. It's basically 180 degrees divided by NTHETA so perhaps I should use that formula rather than hardcoding 0.2. It someone changes NTHETA the current code will break

DannyHindson committed:
Address review comments

Add some additional filters on new getXMinMax function to ignore
monitors and spectra without detectors
Add missing doxygen documentation header
Replace magic 0.2 factor in unit test with a variable instead

RichardWaiteSTFC reviewed (dismissed):
Looks good, really big speed up! Thanks for the clarifications and the additional checks.

martyngigg reviewed (commented):
Could I just ask about the new file [Framework/TestHelpers/src/WORKSPACE2D.ipch](https://github.com/mantidproject/mantid/pull/33545/files#diff-f81bc11b37ea4a4531862d60cb91b8df1dd201937052e51aef73c9a70c140bc2)? As it's binary is it intended to be committed?

DannyHindson commented:
> Could I just ask about the new file [Framework/TestHelpers/src/WORKSPACE2D.ipch](https://github.com/mantidproject/mantid/pull/33545/files#diff-f81bc11b37ea4a4531862d60cb91b8df1dd201937052e51aef73c9a70c140bc2)? As it's binary is it intended to be committed?

No that's not supposed to be there. I'll get rid of it. Seems to be some sort of Visual Studio file.

DannyHindson committed:
Remove stray ipch file

RichardWaiteSTFC reviewed (approved):
.ipch file removed so re-approving - sorry I didn't spot it before!
